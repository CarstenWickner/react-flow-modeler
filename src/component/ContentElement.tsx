import * as React from "react";

import { HorizontalStroke } from "./HorizontalStroke";

export class ContentElement extends React.Component<{
    elementId: string;
    editMenu: (() => React.ReactNode) | undefined;
    onSelect: (elementId: string) => void;
}> {
    onClick = (event: React.MouseEvent): void => {
        const { onSelect, elementId } = this.props;
        onSelect(elementId);
        event.stopPropagation();
    };

    render(): React.ReactNode {
        const { editMenu, children } = this.props;
        return (
            <>
                <div className="stroke-horizontal arrow" />
                <div className="flow-element-wrapper">
                    <div className={`flow-element content-element${editMenu ? " selected" : ""}`} onClick={this.onClick}>
                        {children}
                    </div>
                    {editMenu && editMenu()}
                </div>
                <HorizontalStroke optional />
            </>
        );
    }
}
