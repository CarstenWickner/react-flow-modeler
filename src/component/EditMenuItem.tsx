import * as React from "react";
import { useDrag } from "react-dnd";

import { DraggedLinkContext } from "../types/EditAction";

const disabledDragging: [{ isDragging: boolean }, React.LegacyRef<HTMLSpanElement>] = [{ isDragging: false }, undefined];

export const EditMenuItem: React.FC<{
    options: { className?: string; title?: string };
    defaultClassName: string;
    onClick?: (event: React.MouseEvent) => void;
    dragItem?: DraggedLinkContext;
}> = ({ options, defaultClassName, onClick, dragItem }) => {
    if (!dragItem && !onClick) {
        return null;
    }
    const [{ isDragging }, drag] = dragItem
        ? useDrag({
              item: dragItem,
              collect: (monitor) => ({ isDragging: !!monitor.isDragging() })
          })
        : disabledDragging;
    const baseClassName = (options && options.className) || `menu-item ${defaultClassName}`;
    return (
        <span
            className={`${baseClassName}${isDragging ? " dragging" : ""}`}
            {...{ title: options ? options.title : undefined, ref: drag, onClick }}
        />
    );
};
