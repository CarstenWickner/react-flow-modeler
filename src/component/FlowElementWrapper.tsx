import * as React from "react";
import { useDrop } from "react-dnd";

import { isFlowValid } from "../model/pathValidationUtils";

import { ContentNode, ConvergingGatewayNode, DivergingGatewayNode, ElementType, EndNode, ModelElementExclStart } from "../types/ModelElement";
import { DraggableType, DraggedLinkContext, onLinkDropCallback } from "../types/EditAction";

const isTargetAncestorOfDragItem = (
    originElement: ModelElementExclStart,
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode
): boolean => {
    if (originElement === referenceElement) {
        return true;
    }
    if (originElement.type === ElementType.ConvergingGatewayNode) {
        return originElement.precedingBranches.some((branch) => isTargetAncestorOfDragItem(branch, referenceElement));
    }
    return (
        originElement.precedingElement.type !== ElementType.StartNode && isTargetAncestorOfDragItem(originElement.precedingElement, referenceElement)
    );
};

const isDropValid = (referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode, onDrop: onLinkDropCallback) => (
    dragContext: DraggedLinkContext
): boolean =>
    !isTargetAncestorOfDragItem(dragContext.originElement, referenceElement) &&
    dragContext.originElement.followingElement !== referenceElement &&
    isFlowValid(onDrop(referenceElement, dragContext, true).changedFlow);

const disabledDropping: [{ isOver: boolean; canDrop: false }, React.LegacyRef<HTMLDivElement>] = [{ isOver: false, canDrop: false }, undefined];

export const FlowElementWrapper: React.FC<{
    elementTypeClassName: string;
    referenceElement: ContentNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode;
    editMenu?: (() => React.ReactNode) | undefined;
    onLinkDrop?: onLinkDropCallback | undefined;
    onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
}> = ({ elementTypeClassName, referenceElement, editMenu, onLinkDrop, onClick, children }) => {
    const [{ isOver, canDrop }, drop] = onLinkDrop
        ? useDrop({
              accept: DraggableType.LINK,
              canDrop: isDropValid(referenceElement, onLinkDrop),
              drop: (dragContext: DraggedLinkContext) => onLinkDrop(referenceElement, dragContext),
              collect: (monitor) => ({
                  isOver: !!monitor.isOver(),
                  canDrop: !!monitor.canDrop()
              })
          })
        : disabledDropping;
    return (
        <>
            <div className={`stroke-horizontal arrow${isOver && canDrop ? " can-drop" : ""}`} />
            <div className={`flow-element-wrapper${isOver ? (canDrop ? " can-drop" : " cannot-drop") : ""}`} ref={drop}>
                <div className={`flow-element ${elementTypeClassName}${editMenu ? " selected" : ""}`} onClick={onClick}>
                    {children}
                </div>
                {editMenu && editMenu()}
            </div>
        </>
    );
};
