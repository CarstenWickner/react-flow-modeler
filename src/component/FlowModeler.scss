@import "node_modules/mathsass/dist/math";

.flow-modeler {
    display: grid;
    padding: 2em 1em;
    align-content: stretch;
    justify-content: stretch;

    $strokeColor: #01579b;
    $dropStrokeColor: #00b0ff;
    $strokeStrength: 0.25em;
    $arrowSize: 1.5 * $strokeStrength;
    $strokeLength: 1.75em;
    $elementSizeFactor: 2.5;
    $standardElementSize: $elementSizeFactor * 1em;
    $gatewayBorderSize: sqrt($elementSizeFactor * $elementSizeFactor * 2) * 1em / 2;
    $standardElementColor: #808080;
    $selectionColor: #ffab00;

    & .grid-cell {
        display: flex;
        align-items: center;
        align-content: center;
        min-height: $strokeLength;
        position: relative;
    }
    & .flow-element-wrapper {
        display: flex;
        align-items: center;
        position: relative;
    }

    & .flow-element {
        display: flex;
        position: relative;
        margin: $strokeStrength 0;

        &.start-element {
            border: ($strokeStrength / 2) solid $standardElementColor;
            height: $standardElementSize - $strokeStrength;
            width: $standardElementSize - $strokeStrength;
            border-radius: 50%;
        }
        &.end-element {
            box-shadow: 0 0 0 ($strokeStrength * 0.9) $standardElementColor;
            border: ($strokeStrength * 0.9) solid white;
            background-color: $standardElementColor;
            height: $standardElementSize - (4 * 0.9 * $strokeStrength);
            width: $standardElementSize - (4 * 0.9 * $strokeStrength);
            border-radius: 50%;
            margin: $strokeStrength ($strokeStrength * 0.9);
        }
        &.step-element {
            border: ($strokeStrength / 3) solid $standardElementColor;
            border-radius: 0.25em;
            min-width: $standardElementSize;
            min-height: $standardElementSize / 2;
        }
        &.gateway-element {
            position: relative;
            height: $standardElementSize;
            width: $standardElementSize;
            margin: 0 $strokeStrength / 2;
            z-index: 1;

            display: flex;
            align-items: center;
            justify-content: center;

            &::before {
                content: "";
                position: absolute;
                transform: rotate(45deg);
                border: ($strokeStrength / 2) solid $standardElementColor;
                height: $gatewayBorderSize;
                width: $gatewayBorderSize;
                background-color: transparent;
            }
            &::after {
                content: "X";
                position: absolute;
                font: bold 1.75em Arial, sans-serif;
                color: $standardElementColor;
            }
        }
        &.start-element.selected,
        &.step-element.selected,
        &.gateway-element.selected::before {
            border-color: $selectionColor;
        }
    }
    & .menu {
        position: absolute;
        display: flex;
        flex-direction: column;
        align-items: center;
        left: calc(100% + #{$strokeStrength});
        width: 1em;
        margin: auto;
        border: ($strokeStrength / 3) solid $selectionColor;
        border-radius: 0.25em;
        background-color: white;
        box-shadow: 0.1em 0.1em 0.25em 0.25em rgba(0, 0, 0, 0.25);
        z-index: 1;
        padding: $strokeStrength;

        &-item {
            font-family: Arial, Helvetica, sans-serif;
            line-height: 1.1em;
            cursor: pointer;
    
            &.add-branch::before {
                content: "↳";
            }
            &.add-step::before {
                content: "■";
            }
            &.add-gateway::before {
                content: "◆";
            }
            &.change-next::before {
                content: "⇥";
            }
            &.remove::before {
                content: "×";
            }
        }
    }
    & .gateway-element.diverging + .menu {
        top: ($strokeStrength * 3) + (1.1em / 2);
    }
    & .stroke-horizontal::before,
    & .stroke-vertical::before {
        background-color: $strokeColor;
    }
    &.editable .stroke-horizontal.selected::before {
        background-color: $selectionColor;
    }
    & .stroke-horizontal {
        position: relative;
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        min-width: $strokeLength;

        &.optional {
            min-width: 0;
        }
        &.arrow {
            margin-right: $arrowSize;
    
            &::after {
                content: "";
                height: 0;
                width: 0;
                position: absolute;
                right: 0;
                top: calc(50% - #{$arrowSize});
                margin-right: -$arrowSize;
                border-top: $arrowSize solid transparent;
                border-left: $arrowSize solid $strokeColor;
                border-bottom: $arrowSize solid transparent;
            }
            &.can-drop::after {
                border-left-color: $dropStrokeColor;
            }
        }
        &::before {
            position: absolute;
            content: "";
            height: $strokeStrength;
            width: 100%;
            top: calc(50% - #{$strokeStrength / 2});
        }
        &.can-drop::before {
            background: linear-gradient(90deg, $strokeColor 0%, $dropStrokeColor 62%);
        }
        & > .top-label {
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
            align-content: center;
            padding: 0 0.25em;
            margin-bottom: $strokeStrength / 2;
            min-height: $strokeStrength * 4;
        }
        & > .bottom-spacing {
            display: flex;
            flex-direction: column;
            margin-bottom: $strokeStrength / 2;
        }
        & .clickable-spacing {
            min-height: $strokeStrength * 4;
            max-height: $strokeStrength * 8;
            width: 100%;
        }
    }
    & .stroke-vertical {
        position: relative;
        display: flex;
        flex-direction: column;
        align-self: stretch;
        width: $strokeStrength;
        &::before {
            position: absolute;
            content: "";
            width: $strokeStrength;
        }
        &.top-half::before {
            height: calc(50% + #{$strokeStrength / 2});
            top: 0;
        }
        &.full-height::before {
            height: 100%;
            top: 0;
        }
        &.bottom-half::before {
            height: calc(50% + #{$strokeStrength / 2});
            bottom: 0;
        }
    }
    &.editable {
        & .flow-element,
        & .top-label,
        & .clickable-spacing {
            cursor: context-menu;
        }
        & .flow-element.end-element,
        & .selected.flow-element,
        & .selected .top-label,
        & .selected .clickable-spacing {
            cursor: inherit;
        }
        & .start-element.selected,
        & .start-element:hover,
        & .step-element.selected,
        & .step-element:hover,
        & .gateway-element.selected::before,
        & .gateway-element:hover::before {
            box-shadow: 0 0 0.1em 0.1em rgba(0, 0, 0, 0.25);
        }
        & .flow-element-wrapper.can-drop > .step-element,
        & .flow-element-wrapper.can-drop > .gateway-element::before {
            box-shadow: 0 0 0.125em 0.125em rgba(0, 176, 255, 0.5);
        }
        & .flow-element-wrapper.can-drop > .end-element {
            box-shadow:
                0 0 0 ($strokeStrength * 0.9) $standardElementColor,
                0 0 0.125em ($strokeStrength * 0.9 + 0.125em) rgba(0, 176, 255, 0.5);
        }
        & .dragging {
            color: $dropStrokeColor;
        }
        & .flow-element-wrapper.cannot-drop,
        & .flow-element-wrapper.cannot-drop > .flow-element {
            cursor: no-drop;
        }
    }
}
