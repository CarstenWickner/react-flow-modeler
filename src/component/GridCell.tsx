import * as React from "react";

export const GridCell: React.FunctionComponent<{
    colStartIndex: number;
    colEndIndex?: number;
    rowStartIndex: number;
    rowEndIndex?: number;
}> = ({ colStartIndex, colEndIndex, rowStartIndex, rowEndIndex, children }) => (
    <div
        className="grid-cell"
        style={
            {
                gridArea: `${rowStartIndex} / ${colStartIndex} / ${rowEndIndex || "auto"} / ${colEndIndex || "auto"}`
            } as React.CSSProperties
        }
    >
        {children}
    </div>
);
