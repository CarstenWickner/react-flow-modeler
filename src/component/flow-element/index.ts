export { Start } from "./Start";
export { ContentElement } from "./ContentElement";
export { Gateway } from "./Gateway";
export { GatewayConnector } from "./GatewayConnector";
export { StrokeExtension } from "./StrokeExtension";
export { End } from "./End";
