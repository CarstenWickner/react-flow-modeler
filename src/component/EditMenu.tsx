import * as React from "react";

import { EditMenuItem } from "./EditMenuItem";
import { FlowElementReference } from "../model/FlowElement";
import { addContentElement, addDivergingGateway } from "../model/action/addElement";
import { addBranch } from "../model/action/addBranch";
import { isChangeNextElementAllowed } from "../model/action/changeNextElement";
import { removeElement, isRemoveElementAllowed } from "../model/action/removeElement";

import { SelectableElementType, EditActionResult, DraggableType } from "../types/EditAction";
import { FlowModelerProps, MenuOptions } from "../types/FlowModelerProps";
import { ElementType } from "../types/GridCellData";

const onClickStopPropagation = (event: React.MouseEvent): void => event.stopPropagation();

export class EditMenu extends React.Component<{
    targetType: SelectableElementType;
    referenceElement?: FlowElementReference;
    branchIndex?: number;
    menuOptions?: FlowModelerProps["options"]["editActions"];
    onChange: (change: (originalFlow: FlowModelerProps["flow"]) => EditActionResult) => void;
}> {
    isNextElementReferencedByOthers = (): boolean => {
        const { referenceElement, branchIndex } = this.props;
        if (!referenceElement) {
            return false;
        }
        return referenceElement.getFollowingElements()[branchIndex || 0].getPrecedingElements().length > 1;
    };

    renderMenuItem(
        options: MenuOptions,
        defaultClassName: string,
        onClick?: (event: React.MouseEvent) => void,
        dragType?: DraggableType
    ): React.ReactNode {
        const { targetType, referenceElement, branchIndex } = this.props;
        if (options && options.isActionAllowed && !options.isActionAllowed(targetType, referenceElement, branchIndex)) {
            return null;
        }
        const dragItem =
            dragType === DraggableType.LINK &&
            (targetType === ElementType.Content || targetType == ElementType.ConnectGatewayToElement) &&
            this.isNextElementReferencedByOthers()
                ? { type: dragType, originType: targetType, originElement: referenceElement, originBranchIndex: branchIndex }
                : undefined;
        return <EditMenuItem key={defaultClassName} options={options} defaultClassName={defaultClassName} onClick={onClick} dragItem={dragItem} />;
    }

    onAddContentElementClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addContentElement(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddContentElementItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addFollowingContentElement : undefined, "add-content", this.onAddContentElementClick);
    }

    onAddDivergingGatewayClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            onChange((originalFlow) => addDivergingGateway(originalFlow, targetType, {}, referenceElement, branchIndex));
        }
    };

    renderAddDivergingGatewayItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(
            menuOptions ? menuOptions.addFollowingDivergingGateway : undefined,
            "add-gateway",
            this.onAddDivergingGatewayClick
        );
    }

    onAddDivergingBranchClick = (): void => {
        const { targetType, onChange, referenceElement } = this.props;
        if (targetType === ElementType.GatewayDiverging) {
            onChange((originalFlow) => addBranch(originalFlow, {}, referenceElement));
        }
    };

    renderAddDivergingBranchItem(): React.ReactNode {
        const { targetType, menuOptions } = this.props;
        if (targetType !== ElementType.GatewayDiverging) {
            return null;
        }
        return this.renderMenuItem(menuOptions ? menuOptions.addDivergingBranch : undefined, "add-branch", this.onAddDivergingBranchClick);
    }

    renderChangeNextElementItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isChangeNextElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.changeNextElement : undefined, "change-next", undefined, DraggableType.LINK);
        }
        return null;
    }

    onRemoveClick = (): void => {
        const { targetType, onChange, referenceElement, branchIndex } = this.props;
        if (targetType === ElementType.Content || targetType === ElementType.ConnectGatewayToElement) {
            onChange((originalFlow) => removeElement(originalFlow, targetType, referenceElement, branchIndex));
        }
    };

    renderRemoveItem(): React.ReactNode {
        const { targetType, referenceElement, branchIndex, menuOptions } = this.props;
        if (isRemoveElementAllowed(targetType, referenceElement, branchIndex)) {
            return this.renderMenuItem(menuOptions ? menuOptions.removeElement : undefined, "remove", this.onRemoveClick);
        }
        return null;
    }

    render(): React.ReactNode {
        const menuItems = [
            this.renderAddContentElementItem(),
            this.renderAddDivergingGatewayItem(),
            this.renderAddDivergingBranchItem(),
            this.renderChangeNextElementItem(),
            this.renderRemoveItem()
        ];
        if (menuItems.some((item) => item)) {
            return (
                <div className="menu" onClick={onClickStopPropagation}>
                    {menuItems}
                </div>
            );
        }
        return null;
    }
}
