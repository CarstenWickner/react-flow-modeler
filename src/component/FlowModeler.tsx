import * as PropTypes from "prop-types";
import * as React from "react";
import OutsideClickHandler from "react-outside-click-handler";
import { DndProvider } from "react-dnd";
import Backend from "react-dnd-html5-backend";

import { StepElement } from "./StepElement";
import { EditMenu } from "./EditMenu";
import { FlowElementWrapper } from "./FlowElementWrapper";
import { Gateway } from "./Gateway";
import { GridCell } from "./GridCell";
import { HorizontalStroke, getConnectionClassName } from "./HorizontalStroke";
import { buildRenderData } from "./renderDataUtils";
import { changeNextElement } from "../model/action/changeNextElement";
import {
    StepNode,
    ConvergingGatewayNode,
    DivergingGatewayBranch,
    DivergingGatewayNode,
    ElementType,
    EndNode,
    StartNode
} from "../types/ModelElement";
import { EditActionResult, DraggedLinkContext, SelectableElementType } from "../types/EditAction";
import { FlowModelerProps } from "../types/FlowModelerProps";
import { GridCellData } from "../types/GridCellData";

import "./FlowModeler.scss";

type Selector = {
    type: SelectableElementType;
    id?: string;
    branchIndex?: number;
};

interface FlowModelerState {
    selection: null | Selector;
}

const createSelector = (
    reference: { type: ElementType.StartNode } | StepNode | DivergingGatewayNode | DivergingGatewayBranch | ConvergingGatewayNode
): Selector => {
    if (reference.type === ElementType.StartNode) {
        return reference;
    }
    if (reference.type === ElementType.ConvergingGatewayNode) {
        return { type: reference.type, id: reference.followingElement.type === ElementType.EndNode ? null : reference.followingElement.id };
    }
    if (reference.type === ElementType.DivergingGatewayBranch) {
        return {
            type: reference.type,
            id: reference.precedingElement.id,
            branchIndex: reference.branchIndex
        };
    }
    return { type: reference.type, id: reference.id };
};

export class FlowModeler extends React.Component<FlowModelerProps, FlowModelerState> {
    state: FlowModelerState = { selection: null };

    isTargetSelected(
        reference: { type: ElementType.StartNode } | StepNode | DivergingGatewayNode | DivergingGatewayBranch | ConvergingGatewayNode | null
    ): boolean {
        if (this.state.selection === null) {
            return reference === null;
        }
        if (reference === null) {
            return false;
        }
        const selector = createSelector(reference);
        return (
            this.state.selection.type === selector.type &&
            this.state.selection.id === selector.id &&
            this.state.selection.branchIndex === selector.branchIndex
        );
    }

    onSelect = (
        reference: { type: ElementType.StartNode } | StepNode | DivergingGatewayNode | DivergingGatewayBranch | ConvergingGatewayNode | null
    ): void => {
        const { onChange } = this.props;
        if (onChange && !this.isTargetSelected(reference)) {
            this.setState({
                selection: reference === null ? null : createSelector(reference)
            });
        }
    };

    onClickStart = (event: React.MouseEvent): void => {
        this.onSelect({ type: ElementType.StartNode });
        event.stopPropagation();
    };

    clearSelection = (event: React.MouseEvent): void => {
        this.onSelect(null);
        event.stopPropagation();
    };

    handleOnOutsideClick = (): void => {
        if (this.state.selection !== null) {
            this.onSelect(null);
        }
    };

    handleOnChange = (change: (originalFlow: FlowModelerProps["flow"]) => EditActionResult): void => {
        const { flow, onChange } = this.props;
        const { changedFlow } = change(flow);
        this.onSelect(null);
        onChange({ changedFlow });
    };

    renderEditMenu = (
        reference: StartNode | StepNode | DivergingGatewayNode | DivergingGatewayBranch | ConvergingGatewayNode
    ): undefined | (() => React.ReactNode) => {
        if (!this.isTargetSelected(reference)) {
            return undefined;
        }
        const { editActions } = this.props;
        return (): React.ReactNode => <EditMenu referenceElement={reference} editActions={editActions} onChange={this.handleOnChange} />;
    };

    handleOnLinkDrop = (
        dropTarget: StepNode | DivergingGatewayNode | ConvergingGatewayNode | EndNode,
        dragContext: DraggedLinkContext,
        dryRun?: boolean
    ): EditActionResult | undefined => {
        const { originElement } = dragContext;
        const change = (originalFlow: FlowModelerProps["flow"]): EditActionResult => changeNextElement(originalFlow, dropTarget, originElement);
        if (dryRun) {
            const { flow } = this.props;
            // perform change without calling onChange, i.e. without triggering re-render
            return change(flow);
        }
        this.handleOnChange(change);
        // onChange is expected to provide the changed flow as prop update, i.e. no need to return anything here
        return undefined;
    };

    renderFlowElement(cellData: GridCellData, editable: boolean): React.ReactNode {
        const onLinkDrop = editable ? this.handleOnLinkDrop : undefined;
        switch (cellData.type) {
            case ElementType.StartNode:
                const startEditMenu = editable && this.renderEditMenu(cellData);
                return (
                    <>
                        <div className={`flow-element start-element${startEditMenu ? " selected" : ""}`} onClick={this.onClickStart} />
                        {startEditMenu && startEditMenu()}
                    </>
                );
            case ElementType.StepNode:
                const { renderStep } = this.props;
                return (
                    <StepElement
                        referenceElement={cellData}
                        editMenu={editable ? this.renderEditMenu(cellData) : undefined}
                        onLinkDrop={onLinkDrop}
                        onSelect={this.onSelect}
                    >
                        {renderStep(cellData)}
                    </StepElement>
                );
            case ElementType.DivergingGatewayNode:
                const { renderGatewayConditionType } = this.props;
                return (
                    <Gateway
                        gateway={cellData}
                        editMenu={editable ? this.renderEditMenu(cellData) : undefined}
                        onLinkDrop={onLinkDrop}
                        onSelect={this.onSelect}
                    >
                        {renderGatewayConditionType && renderGatewayConditionType(cellData)}
                    </Gateway>
                );
            case ElementType.ConvergingGatewayNode:
                return (
                    <Gateway
                        gateway={cellData}
                        editMenu={editable ? this.renderEditMenu(cellData) : undefined}
                        onLinkDrop={onLinkDrop}
                        onSelect={this.onSelect}
                    />
                );
            case ElementType.DivergingGatewayBranch:
                const { renderGatewayConditionValue } = this.props;
                return (
                    <HorizontalStroke
                        referenceElement={cellData}
                        editMenu={editable ? this.renderEditMenu(cellData) : undefined}
                        onSelect={this.onSelect}
                    >
                        {renderGatewayConditionValue && renderGatewayConditionValue(cellData)}
                    </HorizontalStroke>
                );
            case ElementType.ConvergingGatewayBranch:
                return (
                    <>
                        <div className="stroke-horizontal" />
                        <div className={`stroke-vertical ${getConnectionClassName(cellData)}`} />
                    </>
                );
            case ElementType.EndNode:
                return <FlowElementWrapper referenceElement={cellData} elementTypeClassName="end-element" onLinkDrop={onLinkDrop} />;
        }
    }

    renderGridCell = (editable: boolean): ((cellData: GridCellData) => React.ReactNode) => {
        const { options } = this.props;
        const verticalAlign = options ? options.verticalAlign : undefined;
        return (cellData: GridCellData): React.ReactNode => {
            const { colStartIndex, colEndIndex, rowStartIndex, rowEndIndex } = cellData;
            return (
                <GridCell
                    colStartIndex={colStartIndex}
                    colEndIndex={colEndIndex}
                    rowStartIndex={verticalAlign === "bottom" && rowEndIndex ? rowEndIndex - 1 : rowStartIndex}
                    rowEndIndex={verticalAlign === "top" || verticalAlign === "bottom" ? undefined : rowEndIndex}
                    key={`${colStartIndex}-${rowStartIndex}`}
                >
                    {this.renderFlowElement(cellData, editable)}
                </GridCell>
            );
        };
    };

    render(): React.ReactElement {
        const { flow, options, onChange } = this.props;
        const { gridCellData, columnCount } = buildRenderData(flow, options && options.verticalAlign === "bottom" ? "bottom" : "top");
        return (
            <OutsideClickHandler onOutsideClick={this.handleOnOutsideClick}>
                <DndProvider backend={Backend}>
                    <div
                        className={`flow-modeler${onChange ? " editable" : ""}`}
                        onClick={onChange ? this.clearSelection : undefined}
                        style={{ gridTemplateColumns: `repeat(${columnCount}, max-content)` }}
                    >
                        {gridCellData.map(this.renderGridCell(!!onChange))}
                    </div>
                </DndProvider>
            </OutsideClickHandler>
        );
    }

    static propTypes = {
        flow: PropTypes.shape({
            firstElementId: PropTypes.string,
            elements: PropTypes.objectOf(
                PropTypes.oneOfType([
                    PropTypes.shape({
                        data: PropTypes.object,
                        nextElementId: PropTypes.string
                    }),
                    PropTypes.shape({
                        data: PropTypes.object,
                        nextElements: PropTypes.arrayOf(
                            PropTypes.shape({
                                id: PropTypes.string.isRequired,
                                conditionData: PropTypes.object.isRequired
                            })
                        ).isRequired
                    })
                ])
            ).isRequired
        }).isRequired,
        options: PropTypes.shape({
            verticalAlign: PropTypes.oneOf(["top", "middle", "bottom"])
        }),
        renderStep: PropTypes.func.isRequired,
        renderGatewayConditionType: PropTypes.func,
        renderGatewayConditionValue: PropTypes.func,
        onChange: PropTypes.func,
        editActions: PropTypes.shape({
            addDivergingBranch: PropTypes.shape({
                className: PropTypes.string,
                title: PropTypes.string,
                isActionAllowed: PropTypes.func,
                getBranchConditionData: PropTypes.func
            }),
            addFollowingStepElement: PropTypes.shape({
                className: PropTypes.string,
                title: PropTypes.string,
                isActionAllowed: PropTypes.func,
                getStepData: PropTypes.func
            }),
            addFollowingDivergingGateway: PropTypes.shape({
                className: PropTypes.string,
                title: PropTypes.string,
                isActionAllowed: PropTypes.func,
                getGatewayData: PropTypes.func,
                getBranchConditionData: PropTypes.func
            }),
            changeNextElement: PropTypes.shape({
                className: PropTypes.string,
                title: PropTypes.string,
                isActionAllowed: PropTypes.func
            }),
            removeElement: PropTypes.shape({
                className: PropTypes.string,
                title: PropTypes.string,
                isActionAllowed: PropTypes.func
            })
        })
    };
}
