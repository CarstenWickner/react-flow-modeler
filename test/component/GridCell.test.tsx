import * as React from "react";
import { shallow } from "enzyme";

import { GridCell } from "../../src/component/GridCell";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const component = shallow(
            <GridCell colStartIndex={1} colEndIndex={3} rowStartIndex={4} rowEndIndex={6}>
                <div>cell content</div>
            </GridCell>
        );
        expect(component).toMatchSnapshot();
    });
    it("with colStartIndex = 2 and no colEndIndex", () => {
        const component = shallow(
            <GridCell colStartIndex={2} rowStartIndex={3} rowEndIndex={5}>
                <div>cell content</div>
            </GridCell>
        );
        expect(component.prop("style")).toBeDefined();
        expect(component.prop("style")["grid-area"]).toEqual("3 / 2 / 5 / auto");
    });
    it("with rowStartIndex = 2 and no rowEndIndex", () => {
        const component = shallow(
            <GridCell colStartIndex={2} colEndIndex={5} rowStartIndex={3}>
                <div>cell content</div>
            </GridCell>
        );
        expect(component.prop("style")).toBeDefined();
        expect(component.prop("style")["grid-area"]).toEqual("3 / 2 / auto / 5");
    });
});
