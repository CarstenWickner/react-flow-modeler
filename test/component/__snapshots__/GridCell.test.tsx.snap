// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="grid-cell"
  style={
    Object {
      "gridArea": "4 / 1 / 6 / 3",
    }
  }
>
  <div>
    cell content
  </div>
</div>
`;
